src/transformStateWithClones.js

```diff
@@ -7,7 +7,56 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopies = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        clearObject(stateCopy);
+
+        break;
+
+      case 'addProperties':
+        addProperties(stateCopy, action.extraData);
+
+        break;
+
+      case 'removeProperties':
+        removeProperties(stateCopy, action.keysToRemove);
+
+        break;
+
+      default:
+        continue;
+    }
+
+    stateCopies.push({ ...stateCopy });
+  }
+
+  return stateCopies;
+}
+
+function clearObject(object) {
+  for (const key in object) {
+    delete object[key];
+  }
+
+  // return object.push(object);
+}
+
+function addProperties(object, array) {
+  Object.assign(object, array);
+
+  // return object.push(object)]);
+}
+
+function removeProperties(object, array) {
```

Comment:
[NAMING] - The parameter `array` in `removeProperties` function is misleading because `keysToRemove` is expected to be an array. A name like `keys` would be more appropriate to indicate that it should be an array of keys to remove from the object.